### PR TITLE
[FIX] base: in case we trigger internally we need a session to print

### PR DESCRIPTION
**Pull request by wtaferner**
As we need to retrieve the assets which are only available as
a user who can actually retrieve them (with a session) and we do not
have any request to take the current active session from, we do simply
create a short lived session for the process and destroy it afterwards.

**Description of the issue/feature this PR addresses:**
It is impossible to have a proper report printed as PDF via execution of `ir.cron` as the session is missing (no request given where we could reuse it). As we have now the possibility to create sessions to authenticate without password, we do so and immediately destroy the flagged session afterwards.

**Current behavior before PR:**
Reports created through planned actions are having no layout applied and therefore are useless

**Desired behavior after PR is merged:**
Reports can be created in the name of a user during session independent cron jobs.

This is a proof of concept and not perfect as it is I guess, especially as there is security involved, but I am happy to discuss and finally fix this nasty issue.

@smetl @mart-e @odony @nim-odoo @ged-odoo 
May I ask you to take a short look from your perspective and give me the hints needed to get this fixed and merged? Thank you. It should be possible IMO to apply even on version 9.

Info: @wt-io-it

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by odony** [comment]
I don't think this is a sound approach for whatever problem you are trying to fix. It's brittle and involves layers (security, session store) that aren't supposed to be manually used in this manner.
Moreover, this does not look like something we want to address in stable versions - it really looks like a wishlist / improvement.

**Comment by wtaferner** [comment]
@odony 
Well, I was already afraid you would say something like that, but basically what is the official supported way? 
And if it is not sound, is it allowed to be used that way from any security related perspective code-wise anywhere? 
If not we should secure the way session work and can be created by now that it can't be used as I did, right? Is it then a security issue? 
Please contact me and close down the PR if you think it is.

Anyway, I was happy, that I could have a session as it is needed for report generation without a direct client user request which I actually fixed.
Otherwise I would need a session to get it right and how am I supposed to get the session as `ir.cron``even if I am executing it as superadmin? 
Three ways...knowing the password (which is even worse) and retrieve it officially requesting locally a new session or saving the last session and retrieve it to reuse (bad as well IMO) or do what I did by doing exactly the same without any request.

Maybe the other devs involved can give me some hints how to fix this any other way or share my view?

Any additional help about the issue (I am fine with a confirmation and some suggestions how we could tackle this now or in the future as it is a huge issue IMO)?

Wishlist? Hmm, depends on the point of view, but anyway, ready to discuss and improve, unrelated if it is classified for master or stable.

**Comment by odony** [comment]
There's no hidden security problem, it just looks like a wrong technical solution, because it involves the wrong layers of the system. Like using direct SQL to bypass ACLs when you could use sudo() and use the API properly. See what I mean?

Instead of trying to create a session when there is none, couldn't you remove the need for a session when printing reports? It's a technical limitation that could be lifted, not a requirement.

**Comment by wtaferner** [comment]
I agree that the issue is how wkhtmltopdf is working in context with Odoo, but I am mostly searching for solutions and want to share/suggest it for the sake of general welfare and moving forward that others are not limited forever as something needs to be changed on the reporting layer and is known for years (if not, I would say wow). Automation of report generation (even in PDF) is nothing to be belittled that is not yet properly working in Odoo.

Honestly, I do not know how I could fix it differently, so let's wait for the expertise of @smetl and @ged-odoo how we could remove the need of having a session to properly render reports in wkhtmltopdf without removing wkhtmltopdf as a reporting engine.

I count on you that there will be another solution worked on and prioritized as you now know that there is a huge issue even though I am fine for now by making happy myself with a solution ;-) Even if it is a monkey patch, which I knew from the start and I even assumed that you won't be happy, but good solutions always need a monkey patch first, especially if you are not the publisher or the maintainer ^^

**Comment by sle-odoo** [comment]
@wtaferner can you share a snippet of code showing the failure? i tried to reproduce it but it's really a pain to create a report from the settings...

afaik the assets route are public, so a session should automatically be created when querying the route.

the idea of passing the session of the current request is to print the report as the user, in order to grant him the same access rights. If you decide to print a report from a cron or a server actions, you have to make sure you only read fields anybody can read. But i suppose you can reproduce the issue with only static content?

**Comment by wtaferner** [comment]
Thank you for taking the time. 

So basically, I wanted to automate this via cron and it is of course wanted that a certain user is executing this (not the system user as the person needs to be subscribed automatically and I do not want any hardcoded ids)

**Method for creating automatically the invoice pdf and send it**
- It is working great as a server action executed by the user
- It is working smoothly if you manually execute the cron job calling this method on an invoice object
- It fails by the assets if you try to execute the cron job as it is supposed to work (without an active request with the cron worker) b/c of the issue fixed

```python
@api.multi
def action_invoice_sent_auto(self):
    template = self.env.ref('account.email_template_edi_invoice', False)
    for invoice in self.filtered(lambda inv: not inv.sent and inv.state in ['open', 'paid']):
        partners = invoice.message_follower_ids.mapped('partner_id')
        if not all([partner.email for partner in partners]):
            raise ValidationError(_("Please check all followers of %s to have an email address!") % (invoice.number))
        if template:
            self.env['mail.thread'].message_post_with_template(
                template.id,
                model='account.invoice',
                res_id=invoice.id,
                composition_mode='comment',
                partner_ids=partners.ids,
            )
            invoice.sent = True
        else:
            raise ValidationError(_("No email template defined for automatically sending %s!") % (invoice.number))
````

**Comment by CakJuice** [comment]
is this PR implemented on ver. 10?

**Comment by wtaferner** [comment]
@sle-odoo 
Reproduction successful with my snippet as requested and delivered? ;-)

**Comment by JKE-be** [comment]
Hello @wtaferner 

Sorry, but after reflexion, I still cannot reproduce your error.

Yesterday, we reproduced it because we have not httpWorker using the shell.
But I retry into a cron, and it works as expected.

 ./odoo-bin -d test11 --workers=4 --db-filter=^test11$

```python
    @api.multi
    def myfunc(self):
        env = self.env
        invoice = env['account.invoice'].search([], order='id desc', limit=1)
        print(invoice)
        invoice.message_post_with_template(
            env.ref('account.email_template_edi_invoice').id,
            model=invoice._name,
            res_id=invoice.id,
            composition_mode='comment',
            partner_ids=[invoice.partner_id],
        )
```

cron call model.myfunc()

Log:
```
2018-06-14 08:55:14,186 6224 INFO test11 werkzeug: 127.0.0.1 - - [14/Jun/2018 08:55:14] "GET /web/content/626-32d72d7/web.report_assets_pdf.0.css HTTP/1.1" 200 -
2018-06-14 08:55:14,189 5679 INFO test11 werkzeug: 127.0.0.1 - - [14/Jun/2018 08:55:14] "GET /web/content/623-610e11c/web.report_assets_common.1.css HTTP/1.1" 200 -
2018-06-14 08:55:14,205 6224 INFO test11 werkzeug: 127.0.0.1 - - [14/Jun/2018 08:55:14] "GET /web/content/622-610e11c/web.report_assets_common.0.css HTTP/1.1" 200 -
2018-06-14 08:55:14,894 6222 INFO test11 odoo.addons.base.ir.ir_actions_report: The PDF report has been generated for records [10].
```
And the layout is correct, assets have been loaded correctly...

Can you check that as discuss yesterday, the db filter is correct and that when wk request the assets, the database is correctled matched by default filter: 

> ... INFO **test11** werkzeug: 127.0.0.1 - - [...

**Comment by wtaferner** [comment]
@JKE-be 
Alright, thank you for your follow-up, so, we got some light into the magic and the issue which is not working by design and we blame the missing Odoo dbfilter usage for the 404 of assets. 

**But what could be the solution here?**
The Odoo dbfilter itself just sucks in terms of retrieval of a final single db...too little options IMO.

The session was the best and cleanest option to inject the db correctly without dbfilter which is missing if there is no request (with identified database) involved and was even implemented at some point by Odoo SA ;-) See the cookie injection/reusage [here](https://github.com/odoo/odoo/blame/3425752eaca236c1e4b4d616aa49b3faa6cfd957/addons/report/models/report.py#L415)...

There seems to be no reason why we would officially need the cookie or as a consequence the session besides the retrieval of the correct database that a db route is taken.

[OCA DB Filter](https://github.com/OCA/server-tools/tree/11.0/dbfilter_from_header) is by the way the reason why a lot of people including me are not using the Odoo dbfilter alone which is limited to exactly one DB per instance running if you want to comply with the single db output (waste of resources if you have test and quality databases running on the same code in production).

As a consequence and proof of concept, I fixed the issue now also another customized and maybe better way to make @odony happier to not misuse the session :stuck_out_tongue:  (of course now we could cut the issue and say it is not a Odoo core issue anymore, but I hope you are interested in improving and make it easier to adopt too) and it is hackish with the session and would be also hackish doing it what I tried below, but is probably the best way to go for Odoo in terms of DB Filter improving or fixing by adding a clear database pointer in the header or any other way...the db parameter is now only working in specific scenarios which is good.

#### Solution for all OCA DB Filter Users before the sending of the wkhtmltopdf command

```python
command_args.extend(['--custom-header', 'X-ODOO-DBFILTER', self._cr.dbname])
command_args.extend(['--custom-header-propagation'])
```

How else could we make sure certain requests out of session scope will be dispatched correctly without faking a session in Odoo to have the correct DB used?

May I ask you to internally discuss a similar approach in Odoo Core and replace the code with the incomplete and probably deprecated --cookie command extension?

A HTTP header can be used without a session to choose on request which database is supposed to handle my request without the limitation of having exactly one DB per instance running and the filter still would apply in case set. 

Also don't forget that a domain might not be accessible outside the browser session (Basic Auth) where we are back to ```report.url``` behind the proxy without a proper session or dbname and the subsequent single db result scenario.

And for whatever it might be good for we could use additionally a secret to allow the direct database request handling if we change the way dbfilter is implemented.

So, basically, this PR can be closed as soon we have a solution tackling the db filter issues or finally is decided to be not important enough or even denied for Odoo Core. This issue might help a lot of people to have a solution or the right hints what is "wrong" maybe in their usage anyway.

Much appreciated to sometimes have better means of communication and collaboration for peers outside of Odoo SA. Thx, @JKE-be, @OCA and @Odoo :+1:

**Comment by wtaferner** [comment]
https://github.com/odoo/odoo/pull/74758 🥳
